File: Problems/Algorithms/1046. Last Stone Weight/last_stone_weight.py
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        n = len(stones)

        while n > 1:
            stones = sorted(stones)
            if stones[-1] == stones[-2]:
                n -= 2
                stones = stones[:-2]
            else:
                temp = stones[-1] - stones[-2]
                stones = stones[:-2]
                stones.append(temp)
                n -= 1

        if n == 0:
            return 0
        else:
            return stones[0]

File: Problems/Algorithms/1046. Last Stone Weight/last_stone_weight.rs
use std::collections::BinaryHeap;

impl Solution {
    pub fn last_stone_weight(stones: Vec<i32>) -> i32 {
        let mut hp = BinaryHeap::from(stones);
        while hp.len() > 1 {
            let last = hp.pop().unwrap();
            let prev = hp.pop().unwrap();
            hp.push(last-prev);
        }
        hp.pop().unwrap_or(0)
    }
}

